src/typecheck/mod.rs

```diff
@@ -1452,30 +1853,49 @@ pub fn unify_rows(
             Err(RowUnifError::MissingRow(ident))
         }
         (AbsType::RowExtend(id, ty, t), r2 @ AbsType::RowExtend(_, _, _)) => {
-            let (ty2, t2_tail) = row_add(state, &id, ty.clone(), TypeWrapper::Concrete(r2))?;
+            let (ty2, t2_tail) = row_add(
+                state,
+                &id,
+                ty.clone(),
+                Closure {
+                    ty: TypeWrapper::Concrete(r2),
+                    env: env1.clone(),
```

Comment:
while all those clones? Can't we reuse the environment `env1` directly?

Reply:
No because `env1` is used later, so using it without a `clone` would move it to a new owner.
It may appear again some unnecessary clones through the code that are over-protections for that kind of problems, but I don't think this is one of them